gmt-4.5.14/src/gshhg/gshhg.1
.TH GSHHG 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
gshhg \- Extract ASCII listings from binary GSHHG data files
.SH SYNOPSIS
\fBgshhg\fP \fIbinaryfile.b\fP [ \fB\-A\fP\fIarea\fP ] [ \fB\-G\fP ] [ \fB\-I\fP\fIid\fP ] [ \fB\-L\fP ] [ \fB\-M\fP ] 
[ \fB\-N\fP\fIlevel\fP ] [ \fB\-Qe\fP|\fBi\fP ] > \fIasciifile.txt\fP
.SH DESCRIPTION
\fBgshhg\fP reads the binary coastline (GSHHS) or political boundary or river (WDBII) files and
extracts an ASCII listing.  It automatically handles byte-swabbing between different
architectures.  Optionally, only segment header info can be displayed.  The output header information has
the format \fIID npoints  hierarchical-level source area f_area west east south north container ancestor\fP,
where hierarchical levels for coastline polygons go from 1 (shoreline) to 4 (lake inside
island inside lake inside land).  Source is either W (World Vector Shoreline) or C (CIA World
Data Bank II); lower case is used if a lake is a river-lake (a portion of a river that is so wide it is
better represented by a closed polygon). The \fIwest east south north\fP is
the enclosing rectangle, \fIarea\fP is the polygon area in km^2 while f_area is the actual area of the
ancestor polygon (at full resolution), \fIcontainer\fP is the ID of the polygon that contains this polygon (-1 if none),
and \fIancestor\fP is the ID of the polygon in the full resolution set that was reduced to yield this
polygon (-1 if full resolution since there is no ancestor).
For river and border data the header is simply
\fIID npoints  hierarchical-level source west east south north\fP.  
For more information about the file formats, see TECHNICAL INFORMATION below.
.TP
\fIbinaryfile.b\fP
GSHHS or WDBII binary data file as distributed with the GSHHS data supplement.  Any of the
5 standard resolutions (full, high, intermediate, low, crude) can be used.
.TP
\fB\-A\fP
Only output polygons whose area equals or exceeds the \fIarea\fP value in km^2 [Default outputs all polygons].
.TP
\fB\-G\fP
Write output that can be imported into GNU Octave or Matlab by ending each segment with a NaN-record.
.TP
\fB\-I\fP
Only output information for the polygon that matches \fIid\fP.
Use \fB\-Ic\fP to get all the continents only [Default outputs all polygons].
.TP
\fB\-L\fP
Only output a listing of polygon or line segment headers [Default outputs headers and data records].
.TP
\fB\-M\fP
Start all header records with the GMT multiple segment indicator '>' [Default uses P for polygons
and L for lines].
.TP
\fB\-N\fP
Only output features whose level matches the given \fIlevel\fP [Default will output all levels].
.TP
\fB\-Q\fP
Control what to do with river-lakes (river sections large enough to be stored as closed polygons).
Use \fB\-Qe\fP to exclude them and \fB\-Qi\fP to exclude everything else instead [Default outputs all polygons].
.SH EXAMPLES
.sp
To convert the entire intermediate GSHHS binary data to ASCII files for Octave/Mathlab, run
.br
.sp
gshhg gshhs_i.b -G > gshhs_i.txt
.br
.sp
To only get a listing of the headers for the river data set at full resolution, try
.br
.sp
gshhg wdb_rivers_f.b -L > riverlisting.txt
.br
.sp
To only extract lakes, excluding river-lakes, from the high resolution file, try
.br
.sp
gshhg gshhs_h.b -Ee -N2 > all_lakes.txt
.SH TECHNICAL INFORMATION
Users who wish to access the GSHHG (GSHHS and WDBII) data directly from their custom programs should
consult the gshhg.c and gshhg.h source code and familiarize themselves with the data format and
how various information flags are packed into a single 4-byte integer.  While we do not maintain
any Octave/Matlab code to read these files we are aware that both Mathworks and IDL have made
such tools available to their users.  However, they tend not to update their code and our file
structure has evolved considerably over time, breaking their code.  Here, some general technical
comments on the binary data files are given.
.br
\fBGSHHS\fP: These files contain completely closed polygons of continents and islands (level 1),
lakes (level 2), islands-in-lakes (level 3) and ponds-in-islands-in-lakes (level 4); a particular
level can be extracted using the \fB\-N\fP option.  Continents are identified as the first 6 polygons
and can be extracted via the \fB\-Ic\fP option.  The IDs for the continents are Eurasia (0),
Africa (1), North America (2), South America (3), Antarctica (4), and Australia (5).
Files are sorted on area from large to small.  There are two sub-groups for level 2: Regular lakes and the
so-called "river-lakes", the latter being sections of a river that are so wide to warrant a polygon
representation.  These river-lakes are flagged in the header (also see \fB\-Q\fP).  All five resolutions
are free of self-intersections.  Areas of all features have been computed using a Lambert azimuthal
equal-area projection centered on the polygon centroids, using WGS-84 as the ellipsoid.  GMT use the
GSHHS as a starting point but then partition the polygons into pieces using a resolution-dependent
binning system; parts of the world are then rebuilt into closed polygons on the fly as needed.
For more information on GSHHS processing, see Wessel and Smith (1996).
.br
\fBWDBII\fP.  These files contain sets of line segments not necessarily in any particular order.  Thus,
it is not possible to extract information pertaining to just one river or one country.  Furthermore,
the 4 lower resolutions derive directly from the full resolution by application of the Douglas-Peucker
algorithm (see gshhg_dp), hence self-intersections are increasingly likely as the resolution is
degraded.  Note that the river-lakes included in GSHHS are also duplicated in the WDBII river files
so that each data set can be a stand-alone representation.  Users who wish to access both data sets
can recognize the river-lakes features by examining the header structure (see the source code for
details); they are also the only closed polygons in the WDBII river file.  There are many levels (classes) in the
river file: River-lakes (0), Permanent major rivers (1), Additional major rivers (2), Additional rivers (3),
Minor rivers (4), Intermittent rivers -- major (6), Intermittent rivers -- additional (7),
Intermittent rivers -- minor (8), Major canals (10), Canals of lesser importance (11), and
Canals -- irrigation type (12).  For the border file there are three levels: National boundaries (1),
Internal domestic boundaries (2), and international maritime boundaries (3).  Individual levels or classes
may be extracted via \fB\-N\fP.
.SH REFERENCES
Douglas, D. H., and T. K. Peucker, 1973, Algorithms for the reduction
of the number of points required to represent a digitized line
of its caricature, \fICan. Cartogr., 10\fP, 112-122.
.br
Gorny, A. J., 1977, \fIWorld Data Bank II General User GuideRep. PB 271869\fP, 10pp,
Central Intelligence Agency, Washington, DC.
.br
Soluri, E. A., and V. A. Woodson, 1990, World Vector Shoreline, \fIInt. Hydrograph. Rev., LXVII(1)\fP, 27-35.
.br
Wessel, P., and W. H. F. Smith, 1996, A global, self-consistent, hierarchical, high-resolution
shoreline database, \fIJ. Geophys. Res., 101(B4)\fP, 8741-8743.
.SH "SEE ALSO"
.IR GMT (1),
.IR gshhg_dp (1)
.IR gshhgtograss (1)


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/gshhg/gshhg.c
/*	$Id: gshhg.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *	Copyright (c) 1996-2015 by P. Wessel and W. H. F. Smith
 *	See LICENSE.TXT file for copying and redistribution conditions.
 *
 * PROGRAM:	gshhg.c
 * AUTHOR:	Paul Wessel (pwessel@hawaii.edu)
 * CREATED:	JAN. 28, 1996
 * PURPOSE:	To extract ASCII data from the binary GSHHG shoreline data
 *		as described in the 1996 Wessel & Smith JGR Data Analysis Note.
 * VERSION:	1-JUL-2014.  For use with GSHHG version 2.3.1
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; version 2 or any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	Contact info: www.soest.hawaii.edu/pwessel */

#include "gmt_gshhg.h"

int main (int argc, char **argv)
{
	double w, e, s, n, area, f_area, lon, lat, scale = 10.0, minarea = 0.0;
	char source, kind[2] = {'P', 'L'}, c = '>', *file = NULL;
	char *name[2] = {"polygon", "line"}, container[8], ancestor[8];
	FILE *fp = NULL;
	int k, line, max_east = 270000000, info, single, error, ID, m, octave = 0, rank = -1;
	uint32_t k32;
#ifdef WORDS_BIGENDIAN
	static const int flip = 0;
#else
	static const int flip = 1;
#endif
	int  OK, level, version, greenwich, river, src, msformat = 0, qmode = 0, first = 1;
	size_t n_read;
	struct GSHHG_HEADER h;
	struct GSHHG_POINT p;

	info = single = error = ID = 0;
	for (k = 1; k < argc; k++) {
		if (argv[k][0] == '-') {	/* Option switch */
			switch (argv[k][1]) {
				case 'A':
					minarea = atof (&argv[k][2]);
					break;
				case 'G':
					octave = 1;
					break;
				case 'L':
					info = 1;
					break;
				case 'M':
					msformat = 1;
					break;
				case 'N':
					rank = atoi (&argv[k][2]);
					break;
				case 'I':
					if (argv[k][2] == 'c')
						single = 2;
					else {
						single = 1;
						ID = atoi (&argv[k][2]);
					}
					break;
				case 'Q':
					if (argv[k][2] == 'e')
						qmode = 1;
					else if (argv[k][2] == 'i')
						qmode = 2;
					else
						error++;
					break;
				default:
					error++;
					break;
			}
		}
		else
			file = argv[k];
	}

	if (!file) error++;

	if (argc == 1 || error) {
		fprintf (stderr, "gshhg %s - ASCII export of GSHHG data\n\n", GSHHG_PROG_VERSION);
		fprintf (stderr, "usage:  gshhg gshhh|wdbii_rivers|wdb_borders_[f|h|i|l|c].b [-A<area>] [-G] [-I<id>] [-L] [-M] [-N<level>] [-Qe|i] > ascii.txt\n\n");
		fprintf (stderr, "\tgshhh|wdbii_rivers|wdb_borders_[f|h|i|l|c].b is a GSHHG file with polygons or lines.\n");
		fprintf (stderr, "\n\tOPTIONS:\n");
		fprintf (stderr, "\t-A will extract only polygons whose area is greater or equal than <area>.\n");
		fprintf (stderr, "\t-G will write '%%' at start of each segment header [P or L] (overwrites -M)\n"
				 "\t   and write 'NaN NaN' after each segment to enable import by GNU Octave or Matlab.\n");
		fprintf (stderr, "\t-I Output data for polygon number <id>.  Use -Ic to get all continent polygons\n");
		fprintf (stderr, "\t   [Default is all polygons].\n");
		fprintf (stderr, "\t-L will only list headers (no data output).\n");
		fprintf (stderr, "\t-M will write '>' at start of each segment header [P or L].\n");
		fprintf (stderr, "\t-N will only output features whose level matches <level> [Default is all levels].\n");
		fprintf (stderr, "\t-Q controls river-lakes.  By default all polygons are output.\n");
		fprintf (stderr, "\t   use -Qe to exclude river-lakes, and -Qi to ONLY get river-lakes.\n");
		exit (EXIT_FAILURE);
	}
	if (!file) {
		fprintf (stderr, "gshhg: No data file given!\n");
		error++;
	}

	if ((fp = fopen (file, "rb")) == NULL ) {
		fprintf (stderr, "gshhg:  Could not find file %s.\n", file);
		exit (EXIT_FAILURE);
	}

	n_read = fread ((void *)&h, (size_t)sizeof (struct GSHHG_HEADER), (size_t)1, fp);

	while (n_read == 1) {
		if (flip) {
			h.id = swabi4 ((unsigned int)h.id);
			h.n  = swabi4 ((unsigned int)h.n);
			h.west  = swabi4 ((unsigned int)h.west);
			h.east  = swabi4 ((unsigned int)h.east);
			h.south = swabi4 ((unsigned int)h.south);
			h.north = swabi4 ((unsigned int)h.north);
			h.area  = swabi4 ((unsigned int)h.area);
			h.area_full  = swabi4 ((unsigned int)h.area_full);
			h.flag  = swabi4 ((unsigned int)h.flag);
			h.container = swabi4 ((unsigned int)h.container);
			h.ancestor  = swabi4 ((unsigned int)h.ancestor);
		}
		level = h.flag & 255;				/* Level is 1-4 */
		version = (h.flag >> 8) & 255;			/* Version is 1-255 */
		if (first) fprintf (stderr, "gshhg %s - Found GSHHG/WDBII version %d in file %s\n", GSHHG_PROG_VERSION, version, file);
		greenwich = (h.flag >> 16) & 3;			/* Greenwich is 0-3 */
		src = (h.flag >> 24) & 1;			/* Greenwich is 0 (WDBII) or 1 (WVS) */
		river = (h.flag >> 25) & 1;			/* River is 0 (not river) or 1 (is river) */
		w = h.west  * GSHHG_SCL;			/* Convert from microdegrees to degrees */
		e = h.east  * GSHHG_SCL;
		s = h.south * GSHHG_SCL;
		n = h.north * GSHHG_SCL;
		source = (level > 4) ? 'A' : ((src == 1) ? 'W' : 'C');		/* Either Antarctica, WVS or CIA (WDBII) pedigree */
		if (river) source = tolower ((int)source);	/* Lower case c means river-lake */
		line = (h.area) ? 0 : 1;			/* Either Polygon (0) or Line (1) (if no area) */
		if (version >= 9) {				/* Variable magnitude for area scale */
			m = h.flag >> 26;
			scale = pow (10.0, (double)m);		/* Area scale */
		}
		area = h.area / scale;				/* Now im km^2 */
		f_area = h.area_full / scale;			/* Now im km^2 */

		OK = ((!single || ((single == 1 && h.id == ID) || (single == 2 && h.id <= 5))) && area >= minarea);	/* Skip if not the one (-I) or too small (-A) */
		if (OK && qmode && ((river && qmode == 1) || (!river && qmode == 2))) OK = 0;	/* Skip if riverlake/not riverlake (-Q) */
		if (OK && rank >= 0 && level != rank) OK = 0;		/* Skip if not the right level (-N) */
		first = 0;
		
		if (!msformat) c = kind[line];
	        if (octave) c = '%';
		if (OK) {
			if (line)
				printf ("%c %6d%8d%3d%2c%10.5f%10.5f%10.5f%10.5f\n", c, h.id, h.n, level, source, w, e, s, n);
			else {
				(h.container == -1) ? sprintf (container, "-") : sprintf (container, "%6d", h.container);
				(h.ancestor == -1) ? sprintf (ancestor, "-") : sprintf (ancestor, "%6d", h.ancestor);
				printf ("%c %6d%8d%2d%2c %.12g %.12g%11.5f%11.5f%10.5f%10.5f %s %s\n", c, h.id, h.n, level, source, area, f_area, w, e, s, n, container, ancestor);
			}
		}

		if (info || !OK) {	/* Skip data, only want headers */
			fseek (fp, (long)(h.n * sizeof(struct GSHHG_POINT)), SEEK_CUR);
		}
		else {
			for (k32 = 0; k32 < h.n; k32++) {

				if (fread ((void *)&p, (size_t)sizeof(struct GSHHG_POINT), (size_t)1, fp) != 1) {
					fprintf (stderr, "gshhg:  Error reading file %s for %s %d, point %d.\n", argv[1], name[line], h.id, k32);
					exit (EXIT_FAILURE);
				}
				if (flip) {
					p.x = swabi4 ((unsigned int)p.x);
					p.y = swabi4 ((unsigned int)p.y);
				}
				lon = p.x * GSHHG_SCL;
				if (((greenwich & 1) && p.x > max_east) || (h.west > 180000000)) lon -= 360.0;
				lat = p.y * GSHHG_SCL;
				printf ("%11.6f%11.6f\n", lon, lat);
			}
			if (octave) printf ("NaN\tNaN\n");
		}
		max_east = 180000000;	/* Only Eurasia needs 270 */
		n_read = fread((void *)&h, (size_t)sizeof (struct GSHHG_HEADER), (size_t)1, fp);
	}
		
	fclose (fp);

	exit (EXIT_SUCCESS);
}
